File: app/src/main/java/com/banuba/sdk/example/offscreen/Camera2Simple.java
package com.banuba.sdk.example.offscreen;

import static android.hardware.camera2.CameraMetadata.LENS_FACING_FRONT;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.ImageFormat;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.Image;
import android.media.ImageReader;
import android.os.Handler;
import android.util.Range;
import android.util.Size;
import android.view.Surface;
import android.view.WindowManager;

import androidx.annotation.NonNull;

import com.banuba.sdk.camera.Facing;
import com.banuba.sdk.internal.utils.CameraUtils;
import com.banuba.sdk.internal.utils.Logger;
import com.banuba.sdk.offscreen.ImageOrientation;

import java.util.Collections;

@SuppressWarnings("WeakerAccess")
public class Camera2Simple {
    private static final String TAG = "Camera2Simple";

    private static final int FIXED_FRAME_RATE = 30;

    private final Handler mHandler;
    private final CameraManager mCameraManager;

    private int mDisplaySurfaceRotation = Surface.ROTATION_0;

    private CameraDevice mCameraDevice;
    private ImageReader mImageReader;

    private CaptureRequest.Builder mPreviewRequestBuilder;
    private CameraCaptureSession mCaptureSession;

    @NonNull
    private final Size mPreferredPreviewSize;
    private Size mPreviewSize;
    private Facing cameraFacing = Facing.FRONT;

    private boolean mIsCameraOpened = false;
    private boolean mCameraOpening = false;
    private int mFacing;
    private int mSensorOrientation;
    private int mDeviceOrientationAngle;
    private final FrameReadyCallback mFrameReadyCallback;

    public interface FrameReadyCallback {
        void onFrameReady(@NonNull Image image, @NonNull ImageOrientation imageOrientation);
    }

    public Camera2Simple(@NonNull Context context, @NonNull FrameReadyCallback frameReadyCallback, @NonNull Size preferredPreviewSize) {
        final WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
        if (windowManager != null) {
            mDisplaySurfaceRotation = windowManager.getDefaultDisplay().getRotation();
        }
        mCameraManager = (CameraManager) context.getSystemService(Context.CAMERA_SERVICE);
        mFrameReadyCallback = frameReadyCallback;
        mPreferredPreviewSize = preferredPreviewSize;
        mPreviewSize = preferredPreviewSize;
        mHandler = new Handler(); // Creating handler here means that all camera events processing in this thread
    }

    private final ImageReader.OnImageAvailableListener mOnImageAvailableListener = this::pushFrame;

    private void pushFrame(ImageReader reader) {
        try {
            final Image image = reader.acquireLatestImage();
            if (image != null) {
                final boolean isRequireMirroring = mFacing == LENS_FACING_FRONT;
                ImageOrientation mEPImageFormat = getImageOrientation(isRequireMirroring);
                mFrameReadyCallback.onFrameReady(image, mEPImageFormat);
            }
            // Image will be closed in OffscreenEffectPlayer

        } catch (Exception e) {
            Logger.wtf("Error while pushing frame", e);
        }
    }

    @SuppressLint("MissingPermission")
    private void openCamera(Facing cameraFacing) {
        if (!mIsCameraOpened) {
            Throwable error = null;
            String usedCameraId = null;

            try {
                if (mCameraManager != null) {
                    for (String cameraId : mCameraManager.getCameraIdList()) {
                        final CameraCharacteristics characteristics = mCameraManager.getCameraCharacteristics(cameraId);
                        final Integer facing = characteristics.get(CameraCharacteristics.LENS_FACING);
                        if (facing != null && facing == cameraFacing.getValue()) {
                            usedCameraId = cameraId;
                            setupCameraCharacteristics(characteristics);
                            break;
                        }
                    }
                }
            } catch (Exception e) {
                Logger.wtf(e);
                error = e;
            }

            if (usedCameraId != null) {
                try {
                    mIsCameraOpened = true;
                    mCameraOpening = true;
                    mCameraManager.openCamera(usedCameraId, mStateCallback, null);
                } catch (Exception e) {
                    mIsCameraOpened = false;
                    mCameraOpening = false;
                    Logger.wtf(e);
                    error = e;
                }
            }

            if (error != null) {
                Logger.wtf(error);
            }
        }
    }


    private void setupCameraCharacteristics(@NonNull CameraCharacteristics characteristics) {
        final StreamConfigurationMap map =
                characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);

        if (map != null) {
            mFacing = getLensFacing(characteristics);
            mSensorOrientation = CameraUtils.getSensorOrientation(characteristics);
            mPreviewSize = CameraUtils.getPreviewSize(characteristics, mPreferredPreviewSize);
        }
    }

    private final CameraDevice.StateCallback mStateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(@NonNull CameraDevice cameraDevice) {
            // This method is called when the camera is opened.  We start camera preview here.
            mCameraDevice = cameraDevice;
            createCameraPreviewSession();
            mCameraOpening = false;
        }

        @Override
        public void onDisconnected(@NonNull CameraDevice cameraDevice) {
            onCameraClosedState(cameraDevice);
            mCameraOpening = false;
        }

        @Override
        public void onError(@NonNull CameraDevice cameraDevice, int error) {
            onCameraClosedState(cameraDevice);
            Logger.wtf((new Exception("Camera error: " + error).fillInStackTrace().toString()));
            mCameraOpening = false;
        }

        @Override
        public void onClosed(@NonNull CameraDevice cameraDevice) {
            onCameraClosedState(cameraDevice);
        }
    };

    private void onCameraClosedState(@NonNull CameraDevice cameraDevice) {
        cameraDevice.close();

        if (mCameraDevice == cameraDevice) {
            mCameraDevice = null;
            mIsCameraOpened = false;
        }
    }

    private void createCameraPreviewSession() {
        try {
            createPreviewRequest();

            // Here, we prepare a CameraCaptureSession for camera preview.

            mCameraDevice.createCaptureSession(
                    Collections.singletonList(mImageReader.getSurface()),
                    new CameraCaptureSession.StateCallback() {
                        @Override
                        public void onConfigured(@NonNull CameraCaptureSession cameraCaptureSession) {
                            // The camera is already closed
                            if (null == mCameraDevice) {
                                return;
                            }

                            // When the session is ready, we start displaying the preview.
                            mCaptureSession = cameraCaptureSession;
                            try {
                                mCaptureSession.setRepeatingRequest(
                                        mPreviewRequestBuilder.build(), null, mHandler
                                );
                            } catch (CameraAccessException e) {
                                Logger.wtf(e);
                            }
                        }

                        @Override
                        public void onConfigureFailed(
                                @NonNull CameraCaptureSession cameraCaptureSession
                        ) {
                            Logger.wtf((new RuntimeException(
                                    "CameraCaptureSession.StateCallback.onConfigureFailed"
                            ).fillInStackTrace()).toString());
                        }
                    },
                    null
            );

        } catch (CameraAccessException e) {
            Logger.wtf(e);
        }
    }

    private void createPreviewRequest() throws CameraAccessException {
        mPreviewRequestBuilder = mCameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);

        if (mImageReader != null) {
            mImageReader.close();
        }

        mImageReader = ImageReader.newInstance(mPreviewSize.getWidth(), mPreviewSize.getHeight(), ImageFormat.YUV_420_888, 5);
        mImageReader.setOnImageAvailableListener(mOnImageAvailableListener, null);

        mPreviewRequestBuilder.addTarget(mImageReader.getSurface());

        mPreviewRequestBuilder.set(
                CaptureRequest.CONTROL_AF_MODE,
                CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_VIDEO
        );

        mPreviewRequestBuilder.set(
                CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
                Range.create(FIXED_FRAME_RATE, FIXED_FRAME_RATE)
        );
    }

    public void closeCamera() {
        mIsCameraOpened = false;
        mHandler.removeCallbacksAndMessages(null);

        final CameraCaptureSession cameraCaptureSession = mCaptureSession;
        if (cameraCaptureSession != null) {
            try {
                cameraCaptureSession.stopRepeating();
            } catch (CameraAccessException | IllegalStateException e) {
                Logger.i(e.getMessage());
            }
            cameraCaptureSession.close();
            mCaptureSession = null;
        }

        if (mCameraDevice != null) {
            mCameraDevice.close();
            mCameraDevice = null;
        }

        final ImageReader reader = mImageReader;
        if (reader != null) {
            reader.setOnImageAvailableListener(null, null);
            // Do not close let this do GC
        }
    }

    private ImageOrientation getImageOrientation(boolean isRequireMirroring) {
        int rotation = 90 * mDisplaySurfaceRotation;
        int deviceOrientationAngle = mDeviceOrientationAngle;

        if (mFacing == CameraCharacteristics.LENS_FACING_BACK) {
            rotation = 360 - rotation;

            if (mDeviceOrientationAngle == 0) {
                deviceOrientationAngle = 180;
            } else if (mDeviceOrientationAngle == 180) {
                deviceOrientationAngle = 0;
            }
        }

        final int imageOrientation = (mSensorOrientation + rotation) % 360;

        return ImageOrientation.getForCamera(imageOrientation, deviceOrientationAngle, mDisplaySurfaceRotation, isRequireMirroring);
    }

    public void switchFacing() {
        if (mCameraOpening) {
            return;
        }

        closeCamera();
        cameraFacing = cameraFacing == Facing.FRONT ? Facing.BACK : Facing.FRONT;
        openCamera(cameraFacing);
    }

    public void applyOrientationAngles(int sensorAngle, int displaySurfaceRotation) {
        mDeviceOrientationAngle = sensorAngle;
        mDisplaySurfaceRotation = displaySurfaceRotation;
    }

    public void openCameraAndStartPreview() {
        openCamera(cameraFacing);
    }

    public static int getLensFacing(@NonNull CameraCharacteristics characteristics) {
        final Integer facingRaw = characteristics.get(CameraCharacteristics.LENS_FACING);
        return facingRaw != null ? facingRaw : LENS_FACING_FRONT;
    }
}

File: app/src/main/java/com/banuba/sdk/example/camera/Camera2Simple.kt
package com.banuba.sdk.example.camera

import android.content.Context
import android.graphics.ImageFormat
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CaptureRequest
import android.hardware.camera2.CameraCaptureSession
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraAccessException
import android.media.ImageReader
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.util.Range
import android.util.Size
import android.view.Surface
import com.banuba.sdk.camera.Facing
import com.banuba.sdk.internal.utils.CameraUtils
import com.banuba.sdk.offscreen.ImageOrientation
import com.banuba.sdk.offscreen.OffscreenEffectPlayer

class Camera2Simple(
    context: Context,
    effectPlayer: OffscreenEffectPlayer,
    private val cameraManager: CameraManager,
    private val preferredPreviewSize: Size
) : ICamera2Simple {

    companion object {
        private const val TAG = "Camera2Simple"

        private const val FIXED_FRAME_RATE = 30
    }

    private val handler = Handler(requireNotNull(Looper.myLooper()))

    private val displaySurfaceRotation = context.display?.rotation ?: Surface.ROTATION_0

    private val rotation = 90 * displaySurfaceRotation

    private var sensorOrientation = 0

    private var previewSize = preferredPreviewSize

    private var isCameraOpened = false

    private var cameraDevice: CameraDevice? = null

    private var previewRequestBuilder: CaptureRequest.Builder? = null

    private var imageReader: ImageReader? = null

    private var captureSession: CameraCaptureSession? = null

    private val cameraFacing = Facing.FRONT

    private val onImageAvailableListener = ImageReader.OnImageAvailableListener { reader ->
        reader.acquireLatestImage()?.let { image ->
            try {
                val imageOrientation = (sensorOrientation + rotation) % 360
                val isRequireMirroring = cameraFacing == Facing.FRONT
                val frameOrientation =
                    ImageOrientation.getForCamera(imageOrientation, 0, displaySurfaceRotation, isRequireMirroring)
                effectPlayer.processImage(image, frameOrientation)
            } catch (e: Exception) {
                Log.e(TAG, "Error while process camera image", e)
            }
        }
    }

    private val stateCallback = object : CameraDevice.StateCallback() {
        override fun onOpened(camera: CameraDevice) {
            cameraDevice = camera
            createCameraPreviewSession()
        }

        override fun onDisconnected(camera: CameraDevice) {
            onCameraCloseState(camera)
        }

        override fun onError(camera: CameraDevice, error: Int) {
            onCameraCloseState(camera)
        }

        override fun onClosed(camera: CameraDevice) {
            onCameraCloseState(camera)
        }
    }

    override fun openCameraAndStartPreview() {
        openCamera()
    }

    override fun stopPreviewAndCloseCamera() {
        closeCamera()
    }

    private fun openCamera() {
        if (!isCameraOpened) {
            try {
                for (cameraId in cameraManager.cameraIdList) {
                    val characteristics = cameraManager.getCameraCharacteristics(cameraId)
                    val facing = characteristics.get(CameraCharacteristics.LENS_FACING)
                    facing?.takeIf { it == cameraFacing.value }?.let {
                        setupCameraCharacteristics(characteristics)
                        cameraManager.openCamera(cameraId, stateCallback, null)
                        isCameraOpened = true
                        return
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error when camera open", e)
            }
        }
    }

    private fun closeCamera() {
        isCameraOpened = false
        handler.removeCallbacksAndMessages(null)
        captureSession?.let {
            try {
                it.stopRepeating()
            } catch (e: Exception) {
                Log.e(TAG, "Error when camera close", e)
            }
            it.close()
            captureSession = null
        }
        cameraDevice?.let {
            it.close()
            cameraDevice = null
        }
        imageReader?.setOnImageAvailableListener(null, null)
    }

    private fun setupCameraCharacteristics(characteristics: CameraCharacteristics) =
        characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)?.let {
            sensorOrientation = CameraUtils.getSensorOrientation(characteristics)
            previewSize = CameraUtils.getPreviewSize(characteristics, preferredPreviewSize)
        }

    private fun createCameraPreviewSession() {
        try {
            createPreviewRequest()
            cameraDevice?.createCaptureSession(
                listOf(imageReader?.surface),
                object : CameraCaptureSession.StateCallback() {
                    override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
                        // The camera is already closed
                        if (cameraDevice == null) return

                        // When the session is ready, we start displaying the preview.
                        captureSession = cameraCaptureSession
                        try {
                            previewRequestBuilder?.build()?.let {
                                cameraCaptureSession.setRepeatingRequest(
                                    it, null, handler
                                )
                            }
                        } catch (e: CameraAccessException) {
                            Log.e(TAG, "Error when configure camera preview session", e)
                        }
                    }

                    override fun onConfigureFailed(
                        cameraCaptureSession: CameraCaptureSession
                    ) {
                        Log.e(TAG, "Error when configure camera preview session")
                    }
                },
                null
            )
        } catch (e: CameraAccessException) {
            Log.e(TAG, "Error when create camera preview session", e)
        }
    }

    @Throws(CameraAccessException::class)
    private fun createPreviewRequest() {
        previewRequestBuilder = cameraDevice?.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
        imageReader?.close()
        imageReader = ImageReader.newInstance(
            previewSize.width, previewSize.height, ImageFormat.YUV_420_888,5
        ).apply {
            setOnImageAvailableListener(onImageAvailableListener, null)
            previewRequestBuilder?.let {
                it.addTarget(surface)
                it.set(
                    CaptureRequest.CONTROL_AE_MODE,
                    CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_VIDEO
                )
                it.set(
                    CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
                    Range.create(FIXED_FRAME_RATE, FIXED_FRAME_RATE)
                )
            }
        }
    }

    private fun onCameraCloseState(camera: CameraDevice) {
        camera.close()
        if (cameraDevice == camera) {
            cameraDevice = null
            isCameraOpened = false
        }
    }
}